include/crocoddyl/core/integrator/rk4.hxx

```diff
@@ -0,0 +1,239 @@
+  d->dli_du[0] = d->differential[0]->Lu;
+
+  d->ddli_ddx[0] = d->differential[0]->Lxx;
+  d->ddli_ddu[0] = d->differential[0]->Luu;
+  d->ddli_dxdu[0] = d->differential[0]->Lxu;
+
+  for (std::size_t i = 1; i < 4; ++i) {
+    differential_->calcDiff(d->differential[i], d->y[i], u);
+    d->dki_dy[i].topRightCorner(nv, nv).diagonal().array() = (Scalar)1;
+    d->dki_dy[i].bottomRows(nv).noalias() = d->differential[i]->Fx;
+
+    d->dyi_dx[i].noalias() = d->dki_dx[i - 1] * rk4_c_[i] * time_step_;
+    differential_->get_state()->JintegrateTransport(x, d->dx_rk4[i], d->dyi_dx[i], second);
+    differential_->get_state()->Jintegrate(x, d->dx_rk4[i], d->dyi_dx[i], d->dyi_dx[i], first, addto);
+    d->dki_dx[i].noalias() = d->dki_dy[i] * d->dyi_dx[i];
+
+    d->dyi_du[i].noalias() = d->dki_du[i - 1] * rk4_c_[i] * time_step_;
+    differential_->get_state()->JintegrateTransport(x, d->dx_rk4[i], d->dyi_du[i], second);
+    d->dki_du[i].noalias() = d->dki_dy[i] * d->dyi_du[i];
+    d->dki_du[i].bottomRows(nv) += d->differential[i]->Fu;
+
+    d->dli_dx[i].noalias() = d->differential[i]->Lx.transpose() * d->dyi_dx[i];
+    d->dli_du[i].noalias() = d->differential[i]->Lu.transpose() + d->differential[i]->Lx.transpose() * d->dyi_du[i];
+    d->ddli_ddx[i] = d->dyi_dx[i].transpose() * d->differential[i]->Lxx * d->dyi_dx[i];
+    d->Luu_partialx[i] = d->differential[i]->Lxu.transpose() * d->dyi_du[i];
+    d->ddli_ddu[i] = d->differential[i]->Luu + d->Luu_partialx[i].transpose() + d->Luu_partialx[i] +
+                     d->dyi_du[i].transpose() * d->differential[i]->Lxx * d->dyi_du[i];
+    d->ddli_dxdu[i] = d->dyi_dx[i].transpose() * d->differential[i]->Lxu +
+                      d->dyi_dx[i].transpose() * d->differential[i]->Lxx * d->dyi_du[i];
+  }
+
+  if (enable_integration_) {
+    d->Fx.noalias() = time_step_ / 6 * (d->dki_dx[0] + 2 * d->dki_dx[1] + 2 * d->dki_dx[2] + d->dki_dx[3]);
+    differential_->get_state()->JintegrateTransport(x, d->dx, d->Fx, second);
+    differential_->get_state()->Jintegrate(x, d->dx, d->Fx, d->Fx, first, addto);
+
+    d->Fu.noalias() = time_step_ / 6 * (d->dki_du[0] + 2 * d->dki_du[1] + 2 * d->dki_du[2] + d->dki_du[3]);
+    differential_->get_state()->JintegrateTransport(x, d->dx, d->Fu, second);
+
+    d->Lx.noalias() = time_step_ / 6 * (d->dli_dx[0] + 2 * d->dli_dx[1] + 2 * d->dli_dx[2] + d->dli_dx[3]);
+    d->Lu.noalias() = time_step_ / 6 * (d->dli_du[0] + 2 * d->dli_du[1] + 2 * d->dli_du[2] + d->dli_du[3]);
+
+    d->Lxx.noalias() = time_step_ / 6 * (d->ddli_ddx[0] + 2 * d->ddli_ddx[1] + 2 * d->ddli_ddx[2] + d->ddli_ddx[3]);
+    d->Luu.noalias() = time_step_ / 6 * (d->ddli_ddu[0] + 2 * d->ddli_ddu[1] + 2 * d->ddli_ddu[2] + d->ddli_ddu[3]);
+    d->Lxu.noalias() =
+        time_step_ / 6 * (d->ddli_dxdu[0] + 2 * d->ddli_dxdu[1] + 2 * d->ddli_dxdu[2] + d->ddli_dxdu[3]);
+  } else {
+    differential_->get_state()->Jintegrate(x, d->dx, d->Fx, d->Fx);
+    d->Fu.setZero();
+    d->Lx = d->differential[0]->Lx;
```

Comment:
Do not forget to add `dt` in the cost integration. To add more explanation, the legacy code was assuming that the cost is discrete time. But this is not good and consistent.

You could use here any integration scheme, and maybe it is interesting to add other options in future. For now just keep the trapezoidal quadrature as done in Euler integrator.

Reply:
I'm not sure I'm getting how the dt has to be added here. Can you point me at the equivalent line of the Euler integrator please?

Reply:
https://github.com/loco-3d/crocoddyl/blob/master/include/crocoddyl/core/integrator/euler.hxx#L117-L121

Reply:
The integration is done in these lines: https://github.com/PepMS/crocoddyl/blob/238d01eae8c0595888231667681eb4c5bc426f7d/include/crocoddyl/core/integrator/rk4.hxx#L156-L176

The ones that you say are inside the `else`clause.

Reply:
The integration is done in these lines: https://github.com/PepMS/crocoddyl/blob/238d01eae8c0595888231667681eb4c5bc426f7d/include/crocoddyl/core/integrator/rk4.hxx#L156-L176

The ones that you say are inside the `else`clause.